### ISSUE TITLE
Bulk Inserts and Collection Constraints

### ISSUE THREAD

**Issue by jrhicks** [techfort/LokiJS#119]
I don't think I'm using indexes correclty.

When merging 10,000+ records from a remote database into a local database, I have to use the id (not the $loki id) to make sure the document is not already there.  (if its there I want to merge instead of update).  To do this effeciently, I'm trying to use Loki indices:

```
let loki = new Loki('testdb');
myCollection = loki.addCollection('myCollection', {indices:['id']})
```

Then on each insert I do

```
localRecord = myCollection.findOne({id: remoteRecord.id})
if( localRecord ) {
  // Merge code
} else {
  myCollection.insert( remoteRecord );  
}
```
When I **disable** the localRecord check and just insert, things are smoking fast!  Only slowed by data fetching/paging.

* http://screencast.com/t/H3wsetchhQ

When I **enable** the findOne check, things slow down a bit:

* http://screencast.com/t/gBJQ49yBJwW

When I hack an index to the side that is a big hash of ids mapped to $loki I get really fast again.

* http://screencast.com/t/V5R0g7hPt

I think there must be something I don't understand about indexes.

Please advise :)

**Comment by obeliskos** [comment]
Interesting... what you are describing is not so much an index but a constraint.  

Yea the problem being that on every insert you are invalidating the index and it needs to be rebuilt for the findOne.

Maybe @seriousme (who is exploring similar value hash lookups) can look into a unique constraint functionality using an approach similar to what you described.  This might be something you can define on the collection and let loki handle that for you or just a checkConstraint(indexName, value) type test function.

Our current indexes are designed for optimal search speeds when the ratio of insert/read is either above or below 50%. if you do a lot of inserts and one search it rebuilts index once.  if you do a single insert and a lot of finds it also rebuilts once.  insert/find/insert/find is worse case scenario we might alleviate with a check constraint functionality.

**Comment by techfort** [comment]
@jrhicks @obeliskos @seriousme this is a discussion we have been having through email and it's good to see that real-life scenarios are highlighting the necessity for optimization in the indexing system.
In the case of @jrhicks 's app, a unique index is ideal. As pointed out by @seriousme , even on sorted data, hashmap lookups will always be faster than binary search.
From this I draw two conclusions:
1. the need to support the "unique" type of index, much like any major DBMS around
2. the need to let the user choose the best index for their application, so to achieve the best performance.

A unique index is a relatively trivial piece of work to implement, my first and obvious idea is to create a hashmap whose keys are the values of the property to be indexed and the values are their position in the data array. I am open to suggestions!

**Comment by obeliskos** [comment]
Sounds good. Do we also keep $loki id or do we trust user would not have changed the constrained property?

**Comment by techfort** [comment]
i'd say we keep it. get(int) will work with $loki, and we could extend get to work like get(var, indexName) => gets a record by index name. Opinions?

**Comment by obeliskos** [comment]
Sounds good to me!

**Comment by jrhicks** [comment]
@techfort You rock! I was able to ditch my hack index, and just check-before-insert a little differently with much better results http://screencast.com/t/JLOAJLPTfj

Basically the only delay is the data fetching/paging!

So given a batch of 1000 data.  I first find all the ones that are already in my db

```javascript
      for (let d of data) {
        localRec = db[collectionName].findOne({id: d.id})
        if (localRec) {
          localRecords[localRec.id] = localRec;
        }
      }
```

Then, when I loop over the inserts I just check the disposable localRecords.

```
        if (localRec[remoteRec.id]) {
          console.log("TODO: Merge");
        } else {
          let r = db[collectionName].insert(remoteRec);
          localRec[r.id] = r;
          loadJob.data.shift();
        }
```

**Comment by techfort** [comment]
@jrhicks woah that's fast :) so how do we go about making this a LokiJS facility for everybody?

**Comment by jrhicks** [comment]
@techfort I think your on to something with constraints!  

In lieu of constraints, maybe a bulkInsert function that returns the records that failed the constraint???

```
notInsertedRecords = myCollection.bulkInsert(records, {uniqueConstraint: ['id']})
```

**Comment by jrhicks** [comment]
@techfort - I think your correct.  The best thing would be to be able to create a constraint, but also have a bulk insert routine that respected the constraint.  I guess its getting kinda late there - no reply necessary.

**Comment by techfort** [comment]
@jrhicks this sounds good to me. @obeliskos you good with that?

**Comment by obeliskos** [comment]
Sounds good

**Comment by jrhicks** [comment]
@techfort I'm timid to update core code.  Maybe after I've been using the product longer.  I do feel interested in helping to contribute to documentation though.  As time permits, I can start a conversation about that in a diferent thread.

**Comment by techfort** [comment]
@jrhicks no problem, I'll take a look at your implementation on the github project you referenced in the post and credit you in logs/commits/etc. This all looks great, the post /video is impressive!

**Comment by seriousme** [comment]
As already discussed by mail there are various options to deal with indices.  (e.g. exact unique, exact non-unique, ordered (which is the current default) etc).

As shown by the initial posting its rather surprising to users that the first query after insert/update takes so long.  So I'd try to fix that as well by doing binary index updates instead of full index rebuild. The only place to do full index rebuild might be on bulk updates where it is explainable that updating the index  a record at the time might be suboptimal compared to updating the index only once per bulk update. (but even then.. ;-))

I think that the additional feature proposed by @jrhicks  to have:
```
notInsertedRecords = myCollection.bulkInsert(records, {uniqueConstraint: ['id']})
```

is a nice addition to that.
However as with traditional databases I'd rather define a uniqueIndex on the DB upfront and then reject the inserts that would break uniqueness. 

e.g. something like:
```
myCollection = loki.addCollection('myCollection', {indices:[{ 'id': 'uniqueIndex' }]})
notInsertedRecords = myCollection.bulkInsert(records)
```
Where notInserted records would be a list of records with an indication why they were rejected.
(e.g. " 'id' not unique"  etc..)

When doing all this its important to keep in mind that all additional functionality comes at a price of adding complexity and code size. @jrhicks already showed that there is a workaround :-)

Its up to @techfort  to decide which direction he wants to take Loki :-)

**Comment by jrhicks** [comment]
I agree with @seriousme, define the constraint upfront and do not offer the option on the bulkInsert.  I think we all agree on this. @techfort

**Comment by jrhicks** [comment]
To be most backwards compatible, I think the default

* insert

* save

Should throw and error if a constrain was violated.  So we may need a pattern to have methods that return errors as apposed to throwing an error.  Ruby uses an exclamation point to distinguish between the two.  Maybe the soft (return errors on failure).  Might be???

* tryInsert

* trySave

That doesn't.  This communicates clearly that the action may fail and return an error.  Just an idea.

**Comment by techfort** [comment]
Yes, upfront index declaration is my preferred way too. Also the addition of a property in the options property is the most backward compatible.
As for errors, I'm all for returning results, let's not forget that Collection emits 'error' events, that's also an option to manage errors.

@seriousme @obeliskos @jrhicks index update (as @seriousme points out) is the way to go on binary indexes. Funnily enough, I implemented a (quite hidden) Key Value store class in Loki ( * ), and I used a binary search to establish the position at which a new element should slot into an existing sorted structure/array. What makes this Binary Search interesting is that it returns whether a value was found or not, and the index at which the value *is* or *should be* if it were to be inserted. So from there you would simply apply a splice at the specified position and off you go, no need to rebuild the index. You can see that BS function here: https://github.com/techfort/LokiJS/blob/master/src/lokijs.js#L3434
This should save us a bit of time when doing inserts/updates.

Bottom line we're good to go with the upfront declaration as far as I'm concerned.

( * ) the key-value store class is only there because keys can be complex structures such as objects or arrays. This is because I thought maybe at some stage you may want to use Loki objects as keys for values.

**Comment by techfort** [comment]
this is good. Strange though, this message didn't turn up on github..

**Comment by obeliskos** [comment]
great post @seriousme... i think you have great understanding of what can be done with this.

Yes @techfort's function sounds like it should do the trick.  The 'unique' part of this should make it simpler (single position).

I'm sure as this is being implemented we will be able to tell if providing a batch method could offer performance boost.

**Comment by techfort** [comment]
@jrhicks @seriousme @obeliskos I'm going ahead on the development of exact and unique constraints, stay tuned and happy Easter!

**Comment by jrhicks** [comment]
@techfort Awesome!  Happy Easter :)

**Comment by techfort** [comment]
@jrhicks @seriousme @obeliskos i have completed a very *drafty* implementation of unique indexes, with inserts and removes working as they should. Haven't benchmarked by() operations yet (by(field, value) is the method to obtain a record straight from the unique index rather than go through the query engine).
Before applying a pass in find() to check if the field is a unique index (and therefore returning the by() result instead of the query result), I would like to address updates, which at the moment are quite tricky (really only in the case of someone trying to update the property associated with a unique index).
just updating.

**Comment by jrhicks** [comment]
@techfort - I guess you would have to consume more space to map loki's id to unique id to see when the unique id was changed.  Is Loki id assured to be uneditable?  Could unique indexes be set to non-editable?  Maybe these wouldn't be indexes, they would be primary keys?  Just sorta brainstorming here

**Comment by seriousme** [comment]
@techfort : although semantically the uniqueness is a constraint, technically its an index in which you try to insert a value, if the insertion fails its a constraint violation.
With regards to update: I would use the old value ( find record by I'd ) to determine what to remove from the index and then proceed in the same way as an add.  I would also abstract the index/constraint into classes so the collection could maintain a hashmap to map props to constraints/indices. That way the constraints/indices are cleanly isolated from the add, update, remove methods.

**Comment by obeliskos** [comment]
I look forward to this feature.... thanks!   

I will try to look over implementation within a few days but i'm sure you and seriousme can make sure it's done right :)

**Comment by techfort** [comment]
@seriousme here's the problem I'm facing: cloning an object is - at the moment - an option. So imagine I have a collection with an element `{ username: 'joe', $loki: 1}` then the index would look like `{ 'joe': <a reference to the original object>' }`
This means that if i change username of the object to 'jack' the original object is changed, so now the index looks like `{ 'joe': <an-object-with-username-jack> }`
I meant to follow exactly the pattern you illustrated, until i realized the current state of affairs makes that impossible, or convoluted and slow (so much for indexing!).

One quick solution would be to place a very obvious disclaimer in the docs, "if you declare *ANY* unique indexes then the *clone* option will activate by default. The objects you deal with are copies of the objects in the collection, and sync-ing is only possible with coll.update(obj)".

As for abstracting - yes absolutely. In fact you had already produced an example of that - I may just very blatantly grab off that - as it also included some of my work so I won't feel too guilty :D

If you or @obeliskos have any further ideas on this let me know.

**Comment by seriousme** [comment]
@techfort: you might be able tot remove 'Joe' from the index, (idx['Joe'] = undefined) and then add Jack to the idx.

Feel free to take my (and your;-)) idx code and use it as you see fit.

In the current setup you either need to link to the array element (not its position in the array!) or store the records $Loki id.

If the idx points to an id then lookup will be a 2 step:
- get id from idx hash
- get record by id as already available with .get

This method takes a little more effort to resolve but makes serialization  a breeze :-)

**Comment by obeliskos** [comment]
So if our id index looks like { arrayIdx : $loki } and your new idx is on username, you have been adding an index for { 'joe' : obj }.  Seems we need an { $loki : 'joe' } index as well?  Either that or set up meta array (or force cloning).  

So 2 index maps per unique constraint, or meta, or cloning? Are those options?

**Comment by techfort** [comment]
@seriousme yeah funny i had it working like that and then i thought it didn't really boost retrieval that much. So i reverted back to pointing directly to the object from the  key in the index. And yes - i also crashed 100mph face-first into linking to the array position and reverted knowing I was the only one to witness that code, but hey - life is about mistakes as much as it is about success so here's me candidly admitting to that blunder :D

@obeliskos I also thought of the two-maps index, should be pretty solid and would remedy force-cloning.

as a note: I think we should *definitely* return clones (without an option) in v2.

**Comment by obeliskos** [comment]
Just curious, have you tried refactoring idIndex from { id : $loki } to { $loki : object }  (yet?)

I just noticed our benchmarks no longer work under node?  I guess because of the refactoring to html benchmarks?   Should we keep two versions of that source, one for node and one for html.  I can see we will be needing to reference and extend benchmarking soon.

**Comment by techfort** [comment]
@obeliskos actually no i haven't. I'll create a separate branch for this as it's going to be too dangerous to play around with in master, then try the refactoring you suggested. I'll also look at those benchmarks.

**Comment by obeliskos** [comment]
Oh.. yea I guess file compatibility issues as well.  I would probably rename benchmark.js to benchmark_html.js and restore old benchmark.js.  I can do that later or let you do it if you prefer differently.

**Comment by techfort** [comment]
@obeliskos actually Dave if you don't mind doing that i'll stay on indexes so we don't trip each other. Thanks!

**Comment by techfort** [comment]
@obeliskos @jrhicks @seriousme i have added a first draft of the unique index abstracted into a class. First tests pass, I am yet to test updates and removes but it works nice for now.

**Comment by jrhicks** [comment]
Thats awesome @techfort :)  You rock.

**Comment by obeliskos** [comment]
I'm showing 1M+ ops/sec :)

**Comment by seriousme** [comment]
I rest my case  ;-)

**Comment by obeliskos** [comment]
well technically techfort rested this case, you just kind of poked around it :)

**Comment by techfort** [comment]
actually - could you re-run those benchmarks? i "released the breaks" on
the by() method (as in, i'm not checking for the existence of a particular
unique index, if you try to call a by() on a non-existent index you get an
error)
should be a good bit 10% faster! my dev machine is a pentium, and got by()
running at 500k ops/s but it's a pretty crappy processor. Any decent
processor (like an i7) should hit 1.1M+ now.

**Comment by obeliskos** [comment]
Ok my yoga2pro (i5) gets 1.1M - 1.2M which is great.  Seeing this evolve has been very enlightening of the specifics of javascript indexing.  Good work @techfort (and @seriousme) :)

**Comment by techfort** [comment]
With the development of #144 , UniqueIndex and SortedIndex, this issue seems to have been split and taken care of in different features. Closing this, feel free to re-open or open separate issue to address whatever has not been addressed by the various indexes.

**Comment by ericmuyser** [comment]
Hey guys where do I find this class / docs or commit on the proper way to do this?

**Comment by aguynamedben** [comment]
Great thread. I'm also curious how bulk inserts worked out. It looks like maybe one way to optimize bulk inserts is to call `.insertOne` with the 2nd parameter set to `true`. Do I need to emit an event after that?

**Comment by aguynamedben** [comment]
I also noticed that if [`Collection.prototype.update`](https://github.com/techfort/LokiJS/blob/bb96365999807aba53e9b8d38ea24acc0832cc2c/src/lokijs.js#L5749) is passed an array, `adaptiveBinaryIndexes` are disabled during the duration of the updates, then `ensureAllIndexes()` is called to update the indexes. Pretty cool.

It doesn't look like that same optimization is applied to [`Collection.prototype.insert`](https://github.com/techfort/LokiJS/blob/bb96365999807aba53e9b8d38ea24acc0832cc2c/src/lokijs.js#L5606) and I'm just curious if there's a specific reason for that.

Thanks for making this library public, it's freaking awesome! 🙏
